File: 923.3-sum-with-multiplicity.py
#
# @lc app=leetcode id=923 lang=python3
#
# [923] 3Sum With Multiplicity
#

# @lc code=start
from typing import List
from collections import Counter
class Solution:
    def threeSumMulti(self, arr: List[int], target: int) -> int:
        # i < j < k
        #arr[i] + arr[j] + arr[k] == target
        #1 1 2 2 2 5 5 
        MOD = 10**9 + 7
        arr.sort()
        ans = 0
        for i in range(len(arr)):
            target2 = target - arr[i]
            j, k = i + 1, len(arr) - 1
            while j < k:
                cur = arr[j] + arr[k]
                if cur < target2:
                    j += 1
                elif cur > target2:
                    k -= 1
                else:
                    if arr[j] == arr[k]:
                        #k - j + 1 is the sum of first and last element
                        #k - j is total number of elements
                        ans += ((k - j + 1) * (k - j) // 2) % MOD
                        break
                    else:
                        left = right = 1
                        while j + 1 < k and arr[j] == arr[j + 1]:
                            j += 1
                            left += 1
                        while k - 1 > j and arr[k] == arr[k - 1]:
                            k -= 1
                            right += 1
                        ans += (left * right) % MOD
                        j += 1
                        k -= 1
        return ans % MOD
# @lc code=end

File: 923.3-sum-with-multiplicity.java
/*
 * @lc app=leetcode id=923 lang=java
 *
 * [923] 3Sum With Multiplicity
 */
import java.util.*;
// @lc code=start
class Solution {
    public int threeSumMulti(int[] arr, int target) {
        Arrays.sort(arr);
        long ans = 0;
        int mod = (int)1e9 + 7;
        int n = arr.length;
        for(int i = 0; i < n - 2; i++){
            int T = target - arr[i];
            int j = i + 1, k = n -1;
            while(j < k){
                if(arr[j] + arr[k] < T) j++;
                else if (arr[j] + arr[k] > T) k--;
                else{
                    if(arr[j] != arr[k]){
                        int left = 1, right = 1;
                        while(j + 1 < k && arr[j] == arr[j + 1]){
                            left++;
                            j++;
                        }
                        while(k - 1 > j && arr[k] == arr[k - 1]){
                            right++;
                            k--;
                        }
                        ans += left * right;
                        ans %= mod;
                        j++;
                        k--;
                    } else {
                        int m = k - j + 1;
                        ans += (m * (m - 1) / 2);
                        ans %= mod;
                        break;
                    }
                }
            }

        
        }
        return (int) ans; 
    }
 }
// @lc code=end

